File: C++_Code/Person.cpp
#include "Person.h"
#include <iostream>
#include <algorithm>
using namespace std;

Person::Person(std::string name, int tier, std::string job, const std::vector<std::string>& themes, const std::vector<std::string>& peopleLike, const std::vector<std::string>& peopleDislike) :
    Tier(tier), Name(name), Job(job)
{
    for(int i = 0; i < peopleLike.size(); i++)
    {
        Favorites.push_back(peopleLike[i]);
    }

    for(int i = 0; i < peopleDislike.size(); i++)
    {
        Dislikes.push_back(peopleDislike[i]);
    }

    for(int i = 0; i < themes.size(); i++)
    {
        Themes.push_back(themes[i]);
    }

    sort(Favorites.begin(), Favorites.end());
    sort(Dislikes.begin(), Dislikes.end());
    sort(Themes.begin(), Themes.end());

}


void Person::addPerson(Person &newPerson)
{
    if(newPerson.isSpare())
    {
        Relationships.insert({newPerson.getName(), 0});
        return;
    }
    
    int score = 0;

    //Check if newPerson is liked
    if (find(Favorites.begin(), Favorites.end(), newPerson.getName()) != Favorites.end())
    {
        if(this->getTier() == 1)
            score += 1000;
        else if(this->getTier() == 2)
            score += 800;
        else
            score += 600;
    }

    //Check if newPerson is disliked
    if (find(Dislikes.begin(), Dislikes.end(), newPerson.getName()) != Dislikes.end())
    {
        if(this->getTier() == 1)
            score -= 1500;
        else if(this->getTier() == 2)
            score -= 1300;
        else
            score -= 1100;
    }
    
    
    //Check if any themes are Common
    vector<string> common (Themes.size() + newPerson.getThemes().size());
    vector<string>::iterator end = set_intersection(Themes.begin(), Themes.end(), newPerson.getThemes().begin(), newPerson.getThemes().end(), common.begin());
    score += (end - common.begin())*5;


    //Check if jobs are common
    if(newPerson.getJob() != Job)
        score += 5;

    Relationships.insert({newPerson.getName(), score});
}

void Person::printRelationships()
{
    for (auto i : Relationships)
        cout << i.first << "    " << i.second << endl;
}

File: Working_Version/Person.py
import collections

class Person:
    def __init__(self, name, tier = 5, job = None, themes = None, peopleLike = None, peopleDislike = None):
        self.Name = name
        self.Tier = tier
        self.Job = job
        self.Themes = themes
        self.Favorites = peopleLike
        self.Dislikes = peopleDislike
        self.Relationships = collections.defaultdict(int)
    


    def addPerson(self, newPerson, noPplAtTable):
        #Check if both people are spares
        if self.isSpare() and newPerson.isSpare():
            self.Relationships[newPerson.getName()] = -10
            return
        
        #Check if one person is a spare
        if self.isSpare() or newPerson.isSpare():
            self.Relationships[newPerson.getName()] = 0
            return
 
        score = 0


        if newPerson.getName() in self.Dislikes:  #Check if newPerson is a Dislike
            if self.getTier() == 1:
                score -= 1500*(noPplAtTable)
            elif self.getTier() == 2:
                score -= 1300*(noPplAtTable)
            else:
                score -= 1100*(noPplAtTable)
        
        if newPerson.getName() in self.Favorites: #Check if newPerson is a favorite
            if self.getTier() == 1:
                score += 1000
            elif self.getTier() == 2:
                score += 800
            else:
                score += 600
        
        
        
        
        #Check for Common Themes
        score += len(set(self.Themes).intersection(set(newPerson.getThemes()))) * 50

        if newPerson.getJob() != self.Job:
            score += 50

        self.Relationships[newPerson.getName()] = score


    def printRelationships(self):
        for key in self.Relationships:
            print(key, ' -> ', self.Relationships[key])


    def getName(self):
        return self.Name
    
    def getTier(self):
        return self.Tier
    
    def getJob(self):
        return self.Job
    
    def getThemes(self):
        return self.Themes
    
    def isSpare(self):
        return self.Tier == 5
    
    def addScore(self, name, scoreAdd):
        self.Relationships[name] += scoreAdd
    
    def getScore(self, name):
        return self.Relationships[name]
    




